Preliminary support for gaussian model

jolars opened the pull request:
This pull request adds preliminary support for the gaussian model in sgdnet. It seems to work properly in a few basic tests. The setup of the packages is this:

`sgdnet()` is called from R and is designed to mimic the functionality of `glmnet::glmnet()`. The input is remolded to work with the interface from scikit-learn whereafter the C++ function `FitModel()` is called. Checks for sparsity in the feature matrix are made on the R side and are then forwarded to the `FitModel()` function which calls `SagaSolver()`; the latter is templated to work with sparse and dense matrices and then does all the work.

I also added a method for `coef()` and have set up architecture for supporting other objective functions.

I have only really tested the package properly against `Ridge()` from scikitlearn but it provides very similar output in the tests that I have run.

jolars committed:
Add Matrix to Imports in DESCRIPTION and rearrange order of items

jolars committed:
Provide RowNormsMax, GetStepSize, NonzerosInRow, and Preprocess

Provide a few utility functions for sgdnet.

* RowNormsMax returns the maximum rowise
squared l2-norm in a matrix.
* GetStepSize returns the step size required
in SAGA given a family and value that
RowNormsMax returns.
* NonzerosInRow returns nonempty indices in
a row given a matrix.
* Preprocess runs the scikit-learn
preprocessing options, scaling and centering
depending on user input.

jolars committed:
Set compile-time constant for tolerance

This is used to check if we need
to rescale and reset the updating system in
SAGA when wscale becomes too small.

jolars committed:
Setup enums for the families to use in switch statements

jolars committed:
Setup registration and linking to Rcpp

jolars committed:
Provide sgdnet R function

This will be the main wrapper for sgdnet
to fit the variety of models that it will
feature.

It currently includes no assertions and
only supports the gaussian family.

jolars committed:
Provide coef.glmnet() method

This method simply returns the coefficients
from the sgdnet object, currently
stored in the a0 and beta slots and
names the intercept.

jolars committed:
Provide LogSumExp and SoftThresholding

* LogSumExp will be used for the multinomial
family and is setup to avoid under- and
overflows.
* SoftThresholding is the proximal operator
for l1-regularization.

jolars committed:
Support Objective class for loss functions

The objective class will be used to house
the different loss functions for sgdnet.
Currently, only the gaussian objective is
supported.

All the classes will at least feature
member functions Loss and Gradient.

jolars committed:
Provide main function FitModel and SagaSolver

* FitModel provides the link to R. Its
only real purpose is to convert the
feature matrix to arma::mat or arma::sp_mat
and pass it along to SagaSolver.
* SagaSolver is templated to support both
sparse and dense feature matrices and
houses all the fitting procedure. In the
future, this should likely be split up
into smaller parts to better support
fitting paths of lambda values.

jolars committed:
Add a simply test to check the output from sgdnet

jolars committed:
Document coef.sgdnet

jolars committed:
Document sgdnet

jolars committed:
Check in autogenerated RcppExports and NAMESPACE

jolars committed:
Fix coef.sgdnet() errors when there are no rownames in input object

jolars committed:
Fix return conversion to sparse matrix.

jolars committed:
Fix error in example.

A matrix should have been created.

jolars committed:
Document the intercept parameter in sgdnet()

jolars committed:
Add namespace call to methods for as() and import methods package

jolars committed:
Rename test-sgdnet to test-gaussian

This is because we probably will want to
have separate tests for teach family.

jolars committed:
Complement tests with sparse feature input test

jolars committed:
Document functions in sgdnet.cpp

Provide internal documentation for LaggedUpdate,
SagaSolver, and FitModel. Only the FitModel
documentation will be accessible via the
R interface as it is the only function exported.

jolars committed:
Add internal keywords for utility functions

This is only really for clarification and has
no effect since these functions are not
exported to R.

jolars committed:
Upload autogenerated RcppExports.

jolars committed:
Import the permeability dataset from AppliedPredictiveModelling

We will use this dataset for gaussian regression.

This commit adds the raw code for obtaining and formatting the data as well as documentation for the dataset. It also uses the dataset in the examples for sgdnet().

jolars committed:
Avoid diving by zero when normalizing

jolars committed:
Avoid dividing by zero when resetting scale

jolars committed:
Fix computation of weights change in testing stopping criterion

jolars committed:
Make computing and returning loss optional

This feature is only there for development purposes.

jolars committed:
Store indices in a field object.

jolars committed:
Transpose feature matrix

Transpose the feature matrix and access samples by column instead of row since the storage type is set by column in armadillo.

jolars committed:
Expand travis and appveyor builds

Test for OSX on travis and for both release and devel on both appveyor and travis. Also compile with valgrind on travis.

jolars committed:
Don't set the seed here, it should work for all cases.

jolars committed:
Add assertions for general input and gaussian family and accompanying tests

Check the standard options, currently using stopifnot. Perhaps we should consider using more descriptive errors via custom if > stop statements instead.

jolars committed:
Transpose x before calling SagaSolver

jolars committed:
Add an easier example in sgdnet() to allow travis to finish the valgrind build

michaelweylandt commented on DESCRIPTION:
```diff
@@ -20,11 +19,15 @@ URL: https://github.com/jolars/sgdnet
 BugReports: https://github.com/jolars/sgdnet/issues
 RoxygenNote: 6.0.1
 Roxygen: list(markdown = TRUE)
+Depends: R (>= 3.5.0)
```
Is this a hard dependency or just we're not testing for anything earlier?

jolars replied:
There's no real reason other than that I'm developing on 3.5.0, which makes it convenient. It looks like we can drop to 3.3.0 without any issues. I have no strong feelings about this.

tdhock replied:
would be good to drop this dep if possible, to allow people with older R to use our pkg

jolars replied:
Done

jolars committed:
Allow OSX builds to fail on travis.ci

RcppArmadillo 8.500 is currently not available on OSX so the build is failing. Downgrading the requirement might let users use the earlier versions, however, which has significantly worse performance for sparse matrices.

michaelweylandt commented on R/sgdnet.R:
```diff
@@ -0,0 +1,103 @@
+#' Fit a Generalized Linear Model with Elasticnet Regularization
```
Elastic Net is two words

jolars replied:
I adopted this from the glmnet documentation, but I agree.

michaelweylandt commented on R/sgdnet.R:
```diff
@@ -0,0 +1,103 @@
+#' @param intercept whether to fit an intercept or not
+#' @param maxit maximum number of effective passes (epochs)
+#' @param standardize whether to standardize `x` or not -- ignored when
+#'   `intercept == TRUE`.
```
Why is this ignored if we have an intercept? We might still want to standardize `X` so that the columns have the same norm (this is often done to put the variables "on the same footing").

jolars replied:
Sorry, that's a mistake in the documentation. It is the other way around. No standardization is done when there is no intercept. This is how they do it in scikit-learn's Ridge. I don't know why though, but I just translated it over for now.

michaelweylandt commented on R/sgdnet.R:
```diff
@@ -0,0 +1,103 @@
+#'   `intercept == TRUE`.
+#' @param thresh tolerance level for termination of the algorithm
+#' @param return_loss whether to compute and return the loss at each outer
+#'   iteration, only added here for debugging purposes.
```
Maybe expose this as via the options mechanism if it's only useful for debugging.

michaelweylandt commented on R/sgdnet.R:
```diff
@@ -0,0 +1,103 @@
+#' @param x input matrix
+#' @param y response variable
+#' @param family reponse type
+#' @param alpha elasticnet mixing parameter
```
Two words

michaelweylandt commented on R/sgdnet.R:
```diff
@@ -0,0 +1,103 @@
+  # Convert sparse x to dgCMatrix class from package Matrix.
+  if (is_sparse <- inherits(x, "sparseMatrix")) {
+    x <- methods::as(x, "CsparseMatrix")
+    x <- methods::as(x, "dgCMatrix")
```
Why the double conversion? Does `as(x, "dgCMatrix")` not automatically work for all `x` classes?

jolars replied:
I borrowed this from glmnet

https://github.com/cran/glmnet/blob/ff79c8962562e278abfaafb98641a7aecf140b23/R/glmnet.R#L91

It seemed strange to me but I figured it was there for a reason. I'll look into it.

jolars replied:
No, there doesn't seem to be any reason for this. I am dropping it.

michaelweylandt commented on R/sgdnet.R:
```diff
@@ -0,0 +1,103 @@
+  # Organize return values
+  a0 <- drop(res$a0)
+  names(a0) <- paste0("s", seq_along(a0) - 1L)
+  beta <- Matrix::Matrix(res$beta)
```
Is the plan to only support scalar `lambda` values or is that only what is implemented so far? If the former, why make this back into a (sparse) matrix?

jolars replied:
This is only what's implemented so far. It is converted because it then matches glmnet, even though we have a dense weights matrix internally. I don't know if glmnet actually uses a sparse implementation for their weights but it might make sense.

michaelweylandt commented on tests/testthat/test-gaussian.R:
```diff
@@ -0,0 +1,20 @@
+context("gaussian regression")
```
You can do more tests here based on what we know about the solution for the lasso. At a minimum, I'd test: 

i) lambda = 0 ==> get back OLS solution
ii) lambda > lambda_max ==> get back all zero solution
iii) the correct solution conditional on knowing the active set
iv) Closed form solution for ridge regression

Ping me if you need help devising these tests

jolars replied:
I agree, this should be added too.

tdhock replied:
agree

michaelweylandt commented on src/utils.h:
```diff
@@ -0,0 +1,148 @@
+    // Center feature matrix
+    x_offset = arma::mean(x, 1);
+    for (arma::uword i = 0; i < n_features; ++i)
+      x.row(i) -= x_offset(i);
```
This worries me a bit - if `x` is sparse and large, this will destroy the sparsity pattern (and blow out memory usage).

jolars replied:
Of course, you are right. So just stick to scaling when features are sparse?

michaelweylandt replied:
Scaling isn't the issue - centering is. 

For now, I think it's safe to throw an error, but long term we can probably support sparse X + centering with clever linear algebra. For example, the centered X, which I'll call X_c, can be written as X_c = X - m 1^T where m is the vector of column means and 1 is the vector of all ones. Then, e.g., we can replace X_c^Ty by `(X - m 1^T)^Ty = X^Ty - 1m^Ty` throughout the algorithm and work with X_c without ever materializing it. 

I don't think that's the right adjustment for our algorithm, but it should be possible to work something similar out. Once you write-up the technical / algorithmic docs, I'll take a look and try to work something out.

jolars replied:
Yes, sorry, that's what I meant. I suggest, however, that we just scale the features when the input is sparse and don't throw any errors.

michaelweylandt replied:
Errr.... I don't love the idea of the standardization being different for dense and sparse X. Even if we document it prominently, I would be really surprised if `sgdnet(X, y)` and `sgdnet(as.matrix(X), y)` gave different results...

jolars replied:
Well, if I am not mistaken, the only differences that emerge will result from differences in the relative size of the penalty. Since we are rescaling, we are otherwise going to get back the same coefficients. In any case, there are always going to be some discrepancies because of the differences in how the intercept is handled between the dense and sparse implementations.

```r
x <- Matrix::rsparsematrix(10000, 2, density = 0.5)
x[, 2] <- x[, 2]*100
x[x[, 2] != 0, 2] <- x[x[, 2] != 0, 2] + 10000
y <- x[, 1] + x[, 2]/100

set.seed(0)
coef1 <- coef(sgdnet(x, y, alpha = 0, lambda = 0))
coef2 <- coef(sgdnet(as.matrix(x), y, alpha = 0, lambda = 0))
cbind(coef1, coef2)

# 3 x 2 Matrix of class "dgeMatrix"
# s0          s0
# (Intercept) 0.185315314 0.069078897
# 1           1.005611802 0.993138237
# 2           0.009968166 0.009986274
```

It feels like there should be a way around this. glmnet returns equivalent results for sparse and dense matrices alike. I'll look into how they standardize their features in the sparse implementation.

michaelweylandt replied:
Re: glmnet -- They do a variant of the linear algebra trick I described, though it's a bit simpler since it's a coordinate descent method: x_{sc}^Ty = (x/s - m)^Ty = x^Ty/s - my

Can you elaborate a bit more on this?

    there are always going to be some discrepancies because
    of the differences in how the intercept is handled between
    the dense and sparse implementations.

This doesn't seem right at all. If sklearn does not give the same results for an otherwise equivalent sparse and dense X, I'd vote strongly for diverging from their behavior. I would expect `coef1` and `coef2` to be in every way equivalent (other than computational speed) in your example above and I imagine our users will also.

jolars replied:
[Someone else already did the legwork on this](https://stackoverflow.com/questions/46835181/sklearn-perceptron-behaving-differently-for-sparse-and-dense) but they basically use a slower learning rate for the intercept in the sparse implementation to avoid intercept oscillation.

michaelweylandt replied:
That's interesting - ok, let's table this for now and possibly come back to it later. It's not worth holding up other development right now. 

Can you open an issue so we don't forget about it?

jolars replied:
Yes, good point. It would not be terribly hard to run some simulations and see how this plays out in our implementation.

[The issue has been filed](https://github.com/jolars/sgdnet/issues/2).

michaelweylandt commented on src/utils.h:
```diff
@@ -0,0 +1,148 @@
+                arma::mat&    y,
+                const bool    normalize,
+                const bool    fit_intercept,
+                arma::vec&    x_offset,
```
Can we use a different term than "offset"? 

"Offset" has a statistical meaning (fixed term added to linear predictor in GLM) and IIRC we plan on supporting that sense of "offset" eventually.

jolars replied:
Sure. Center? Mean?

michaelweylandt replied:
Either works. When we add observation weights eventually, this might need to be a weighted mean (not sure what `glmnet` does), so "center" might be less ambiguous.

michaelweylandt commented on src/math.h:
```diff
@@ -0,0 +1,45 @@
+inline double SoftThresholding(const double x,
+                               const double shrinkage) {
+  return std::max(x - shrinkage, 0.0) - std::max(-x - shrinkage, 0.0);
+}
```
Should we make a "prox" object from which this inherits? That will make it easier to add other proximal operators (SCAD/MCP/group lasso) eventually

jolars replied:
Good idea. I have been thinking of something along those lines already.

michaelweylandt commented on src/objectives.h:
```diff
@@ -0,0 +1,24 @@
+
+  class Gaussian : public Objective {
+  public:
+    double Loss(const arma::mat& p, const arma::mat& y) {
```
Can we call this "mu" to match the statistical literature? (mu = conditional mean = E[Y|X])

jolars replied:
Hm, is mu really used that often for conditional mean? I suggest just spelling out `prediction` or `conditonal_mean` instead.

michaelweylandt replied:
Fine with me -- I'm always happy with more descriptive names. ("p" just has another meaning as number of features) Both names work well, though I'm slightly biased towards "conditional mean" as "prediction" carries a few more connotations.

michaelweylandt reviewed (commented):
I think the design looks good overall and the use of templating + sparse X support is nice (though there's one spot where I'm concerned you'll loose the sparsity pattern) I had a bit of trouble following the algorithm at one or two points since it doesn't follow the main SAGA paper explicitly. [BTW - Is the lagged update the algorithm from Appendix D?] 

It's probably worth putting in a vignette / technical note which explains the algorithm in painstaking detail before the summer is over. 

For the Gaussian case, we can add some tests based on what we know from the math (+ some noise for the stochastic-ness). We won't be able to do the same for other losses, but it will be a start. 

Needs more docs eventually, but those can be added later.

jolars committed:
Separate Elastic and Net in title

jolars committed:
Correct TRUE to FALSE in documentation

This reflects the actual functionality as it stands now.

jolars committed:
Separate elastic and net in elasticnet

jolars committed:
Merge branch 'gaussian_model' of github.com:jolars/sgdnet into gaussian_model

tdhock commented on R/sgdnet.R:
```diff
@@ -0,0 +1,103 @@
+#' @param maxit maximum number of effective passes (epochs)
+#' @param standardize whether to standardize `x` or not -- ignored when
+#'   `intercept == FALSE`.
+#' @param thresh tolerance level for termination of the algorithm
```
would be nice to have details about how this tolerance is used -- when do we stop? using the difference between gradient norm, objective function values, etc?

tdhock commented on src/constants.h:
```diff
@@ -0,0 +1,25 @@
+
+namespace sgdnet {
+  constexpr double SMALL =
+    std::pow(std::numeric_limits<double>::epsilon(), 0.75);
```
please add a comment about what this is used for and why you need e^0.75 -- its not obvious to me

tdhock commented on tests/testthat/test-gaussian.R:
```diff
@@ -0,0 +1,20 @@
+context("gaussian regression")
+
+test_that("gaussian regression with dense features work", {
+  x <- matrix(rnorm(30), 10, 3)
```
would be great if you could use real data rather than (or in addition to) simulated data for the tests

tdhock reviewed (approved):
looks good to me, please merge after addressing my minor comments

jolars committed:
Do not center sparse features

We also pass is_sparse along to SagaSolver.

jolars committed:
Move predictions to separate PredictSample function

The reason for this is that we want to make it easier to profile the code in the future.

jolars committed:
Rename x_offset and y_offset to *_center

jolars committed:
Relax R version requirement to 3.3.0

Also make the necessary changes to build on oldrel on travis.ci and appveyor.

jolars committed:
Rename p to prediction in Objective class

jolars committed:
Reformat according to googles C++ style guide

jolars committed:
Supply special class for proximal operators

In addition, rename bool prox to nontrivial_prox to separate this flag from the new class object.

jolars committed:
Provide a more detailed description of SoftThreshold

jolars committed:
Drop double conversion

This was only included to start with because it is what glmnet does. I could not find any reason for it, however.

jolars committed:
Document how thresh and tolerance is used as stopping criteria

jolars committed:
Add @noRd keywords to roxygen comments to avoid generating documentation from c++ code

jolars committed:
Document tolerance

jolars committed:
Flesh out documentation for proximal operators and use roxygen-styled comments

jolars committed:
Remove unnecessary brackets

jolars committed:
Move initialization of intercept_correction out of loop

jolars committed:
Fix intercept fitting when features are sparse

jolars committed:
Add a comment and TODO regarding SMALL constant

jolars committed:
Add copyright notice at top

jolars committed:
Expand tests

Expand tests to now cover

* Comparing theoretical ridge regression results withthose from sgdnet
* Comparing results from OLS regression.
* Comparing output from dense and sparse features.

We also now use real data sets instead of only simulated data.

jolars committed:
Use an option sgdnet.debug to flag if we want debugging information such as loss

jolars committed:
Do not indent as per Google C++ style guide

jolars committed:
Provide package-level documentation with information on options

jolars committed:
Use real data in example

michaelweylandt commented on R/sgdnet.R:
```diff
@@ -9,19 +25,21 @@
+#' @param thresh tolerance level for termination of the algorithm. The
+#'   algorithm terminates when
+#'   \deqn{
+#'     \frac{\mathrm{change~in~weights}}{\mathrm{weights}} < \mathrm{thresh}.
```
I'd vote to call these "coefficients" since weights is already used to refer to i) observation weights; and ii) feature weights. 

It might be clearer to write this as something like

\frac{\|\beta^{(t)} - \beta^{(t-1)}\|_{\infty}}{\|\beta^{(t)}\|_{\infty}} < \mathrm{thresh}

jolars replied:
Hm, in that case I think we should probably switch from using weights for feature weights altogether and simply stick to coefficients or betas. It would only be confusing to use beta or coefficient here but then use weights in the internal API.

michaelweylandt replied:
I'd agree, but maybe wait until you're done porting the basics from sklearn to make the initial port easy? Once that's done, a big rename should be pretty straightforward.

michaelweylandt commented on src/prox.h:
```diff
@@ -0,0 +1,35 @@
+//' @keywords internal
+class Prox {
+public:
+  virtual double Evaluate(const double x, const double shrinkage) = 0;
```
Might be cleaner to use `operator()` here instead. Then you get something like

```
Prox prox = SoftThreshold()
x = SoftThreshold(x, lambda); 
```

jolars replied:
Hm, I'm not sure how to do this with the abstract base class i am using. I am only passing a pointer to `LaggedUpdates()`.

michaelweylandt replied:
I'm still a bit unclear on the use of `LaggedUpdates` until I see the math, but this probably isn't important if you have something that already works.

jolars committed:
Do not store a field of all the vector indices

Also, we rename current_nonzero_indices to nonzero_indices and remove some unnecessary subsetting in LaggedUpdate() when we are accessing all the elements anyway.

jolars commented:
I am merging this one now. I think that most of the issues should be solved, apart from constructing a test for the active set as you suggested @michaelweylandt
